### ISSUE TITLE
SonarQube task fails if FxCop rules are enabled

### ISSUE THREAD

**Issue by pascalberger** [Microsoft/vso-agent-tasks#972]
I tried to enable some FxCop rules in the Sonar Quality Profile. Build fails afterwards in the Finishing Sonar task with the following error:

```
2015-12-11T12:59:00.0251167Z ##[error]Caused by: java.lang.IllegalArgumentException: The property "sonar.cs.fxcop.assembly" must be set and the project must have been built to execute FxCop rules. This property can be automatically set by the Analysis Bootstrapper for Visual Studio Projects plugin, see: http://docs.codehaus.org/x/TAA1Dg.If you wish to skip the analysis of not built projects, set the property "sonar.visualstudio.skipIfNotBuilt".
```

Do I need to setup something to get this working? I assumed that FxCop rules are native supported and no longer require special configuration.

Sonar Version is 5.1.2 with C# plugin 4.3

**Comment by bgavrilMS** [comment]
Hi Pascal,

What version of the c# plugin are you using?

Bogdan
________________________________

**Comment by pascalberger** [comment]
Latest: C# Plugin 4.3

**Comment by bgavrilMS** [comment]
Hi @pascalberger, 

There are 2 potential causes for this. 

1. You are not actually running FxCop on your agent. This may happen if you do not have Visual Studio installed on the agent (as FxCop comes in the box with VS). It can also happen if you are using msbuild 4.0 - we only support msbuild 12 and 14. It may also happen if you customized the build in a way that affects our customizations - have you modified build targets etc. ?

To recap: Are you using a hosted agent? If not, do you have VS installed? What version of msbuild do you use? 

2. The property "sonar.cs.fxcop.assembly" relates to the old VS Bootstrapper plugin, which is not supported. I wonder if you are passing any advanced settings to the analysis, like sonar.visualstudio.enable=true ?

**Comment by pascalberger** [comment]
Thanks @bgavrilMS. I'm using msbuild 4.0, but only because otherwise Sonar won't work at all. See #793. Any idea how to use the NuGet Packaging task with the `-build` flag while running Sonar without having to use `-MSBuildVersion 4`?

**Comment by bgavrilMS** [comment]
Adding @duncanpMS who suggested the first workaround

**Comment by duncanpMS** [comment]
@pascalberger Can you try splitting the analysis and packaging steps to see if that makes a difference? i.e. a build definition something like:
* SonarQube analysis - Begin
* MSBuild
* SonarQube analysis - End
* NuGet package

It still looks like the most likely root cause is the that NuGet packaging step isn't redirecting MSBuild binding requests correctly, as discussed in #793.

Also, how/where exactly are you specifying the  `-MSBuildVersion 4` option to the `nuget pack` command? I haven't been able to repro this locally; `-MSBuildVersion` doesn't seem to be a supported option for the NuGet command line.

**Comment by pascalberger** [comment]
@duncanpMS `MSBuildVersion` is a supported option of the `nuget pack` command. See [here](https://docs.nuget.org/consume/command-line-reference#user-content-pack-command). I'm passing it to the NuGet Arguments of the NuGet Packager (Full arguments are `-Version $(GitVersion.NuGetVersion) -Build -MSBuildVersion 4`).

Also this worked fine once without the `MSBuildVersion` flag as described in #793. Something must have changed to require it to use MSBuild 4.

It works if I only build it inside the Sonar Task and package it outside.

**Comment by duncanpMS** [comment]
Thanks @pascalberger. I've been able to reproduce the issue.

The root cause is that NuGet.exe isn't redirecting bind requests for earlier versions of MSBuild. I've created issue #988 to track that issue. I've suggested an alternative workaround in the new issue (i.e. manually adding a *nuget.exe.config* file on the hosted agent).

The SonarQube tasks are working as expected - you need to use MSBuild 12.0 or MSBuild 14.0 to perform the analysis. on the agent. FYI the message being produced by the newest version of the C# plugin if no *ProjectInfo.xml* files are found has been improved so that it explicitly says that MSBuild 12.0 or later are required.

Consequently, I propose closing this issue and #793 since the underlying issue is now captured in #988. Do you have any objection to this?

**Comment by pascalberger** [comment]
Thanks @duncanpMS! I'm fine with closing this one and proceed in #988.

**Comment by duncanpMS** [comment]
Thanks @pascalberger. FYI I've made a correction to my comment on this page and in #988: the workaround of supplying a *nuget.exe.config* is of course only possible on an on-prem agent, not a hosted agent.
